instrumentation/jmx-metrics/javaagent/README.md

```diff
@@ -221,6 +221,67 @@ Thus, the above definitions will create several metrics, named `my.kafka.streams
 
 The metric descriptions will remain undefined, unless they are provided by the queried MBeans.
 
+### State Metrics
+
+Some JMX attributes expose current state as a non-numeric MBean attribute, in order to capture those as metrics it is recommended to use the special `state` metric type.
+For example, with Tomcat connector, the `Catalina:type=Connector,port=*` MBean has `stateName` (of type `String`), we can define the following rule:
+
+```yaml
+---
+rules:
+  - bean: Catalina:type=Connector,port=*
+    mapping:
+      stateName:
+        type: state
+        metric: tomcat.connector
+        metricAttribute:
+          port: param(port)
+          connector_state:
+            ok: STARTED
+            failed: [STOPPED,FAILED]
+            degraded: _
+```
+
+For a given value of `port`, let's say `8080` This will capture the `tomcat.connector.state` metric of type `updowncounter` with value `0` or `1` and the `state` metric attribute will have a value in [`ok`,`failed`,`degraded`].
+For every sample, 3 metrics will be captured for each value of `state` depending on the value of `stateName`:
+
+When `stateName` = `STARTED`, we have:
+
+- `tomcat.connector` value = `1`, attributes `port` = `8080` and `connector_state` = `ok`
+- `tomcat.connector` value = `0`, attributes `port` = `8080` and `connector_state` = `failed`
+- `tomcat.connector` value = `0`, attributes `port` = `8080` and `connector_state` = `degraded`
+
+When `stateName` = `STOPPED` or `FAILED`, we have:
+
+- `tomcat.connector` value = `0`, attributes `port` = `8080` and `connector_state` = `ok`
+- `tomcat.connector` value = `1`, attributes `port` = `8080` and `connector_state` = `failed`
+- `tomcat.connector` value = `0`, attributes `port` = `8080` and `connector_state` = `degraded`
+
+For other values of `stateName`, we have:
+
+- `tomcat.connector` value = `0`, attributes `port` = `8080` and `connector_state` = `ok`
+- `tomcat.connector` value = `0`, attributes `port` = `8080` and `connector_state` = `failed`
+- `tomcat.connector` value = `1`, attributes `port` = `8080` and `connector_state` = `degraded`
+
+Each state key can be mapped to one or more values of the MBean attribute using:
+- a string literal or a string array
+- a `_` character to provide default option and avoid enumerating all values.
+
+Exactly one `_` value must be present in the mapping to ensure all possible values of the MBean attribute can be mapped to a state key.
```

Comment:
I don't fully understand why this is a requirement. If there are only certain states that you are interested in, why not allow to only specify those?

Reply:
This is not a requirement, but allows to ensure that state metrics are always captured consistently, this is more best practice than a technical requirement.

Doing so means that at a given time, the sum of all samples is equal to 1, which translates to "the state of the thing we observe is known at all times". I think it's worth doing that for now unless we have good examples where it can be problematic.

The example I have in mind when adding this restriction is again probably biased by current focus on Tomcat state, so I'll elaborate a bit:

Let's say we monitor 3 tomcat servers, each of them with a single connector and a state metric to get the connector state with `on|off` values. When consuming those metrics to create a dashboard:

Scenario 1:
- default set to `off` to indicate something is not right.
- creating a linear chart (sum) with breakdown on state will always sum to 3, and the breakdown provides an obvious clue when something is not in `on` state.
- creating a pie-chart with the latest value with breakdown on state also provides a good representation of the 3 tomcat instances.
- the end-user does not even need to know how many tomcat instances exist, you get information from the ratio of `on|off` in the charts and then can investigate what could be wrong if needed.
- creating an extra dedicated `unknown` state could be used to gather those states as they might represent a gap in the metrics mapping.

Scenario 2:
- state value `on` when connector is `RUNNING`, `off` when connector is `STOPPED`, other possible values are not mapped.
- creating a linear chart (sum) with breakdown on state will sum to 3 when the state is either `on` or `off`, but will go down whenever an un-mapped state is present
- the end-user needs to know that the "usual value is 3", and that "when it goes down to less than 3 something is unknown and maybe we have a gap in our mapping"
- it is possible to trigger an alert by monitoring when the value goes down, but this is definitely more complex and depends mostly on the backend and metrics processing.
- with some visualizations it's even more tricky, for example when creating a pie-chart of the latest reported values to represent the "current state", if not all instances can be represented because their state is unknown then they will be ignored. In an extreme case, if only a single instance is in `on` state and the two others are in an unknown state then the pie-chart could be misleading representing "100% is running and it's fine", when in practice it's "only 1/3 fine".

Given that we usually need to monitor things that are working properly and even more the ones that not working properly (which are not in an expected state), having this as a requirement makes it harder to shoot yourself in the foot by missing a state whose value comes from an implementation detail of the monitored product.